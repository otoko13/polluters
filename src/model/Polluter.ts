export interface IPolluterArgs {
  rank: number;
  name: string;
  longitude: number;
  latitude: number;
  globalEmissions: string;
  ownership: string;
  hq: string;
  ceo: string | undefined;
  ceoAnnualPay: string | undefined;
  fossilFuelProduction: string | undefined;
  investmentInRenewables: string | undefined;
  globalEmissions1965to2017: string | undefined;
  projectedIncreaseInProduction2018to2030: string | undefined;
  projectedEmissions2018to2030: string | undefined;
  environmentalDisaster: string | undefined;
  environmentalScandal: string | undefined;
  futureProjects: string | undefined;
  revenue: string | undefined;
}

export enum EOwnership{
  State,
  Shareholders,
  StateAndShareholders,
  Unknown,
}

export interface IPolluter extends Polluter {}

const NO_INFORMATTON = 'No information';

class Polluter {
  readonly name: string;
  readonly rank: number;
  readonly longitude: number;
  readonly latitude: number;
  readonly globalEmissions: string;
  readonly ownership: EOwnership;
  readonly hq: string;
  readonly ceo: string;
  readonly annualCeoPay: number | undefined;
  readonly fossilFuelProduction: string;
  readonly investmentInRenewables: string;
  readonly globalEmissions1965to2017: string;
  readonly projectedIncreaseInProduction2018to2030: string;
  readonly projectedEmissions2018to2030: string;
  readonly environmentalDisaster: string | undefined;
  readonly environmentalScandal: string | undefined;
  readonly futureProjects: string;
  readonly revenue: string;

  private static getOwnership(ownership: string) {
    const isState = ownership.toLowerCase().indexOf('state') > -1;
    const isShareholder = ownership.toLowerCase().indexOf('shareholder') > -1;
    if (isState && isShareholder) {
      return EOwnership.StateAndShareholders;
    }
    if (isState) {
      return EOwnership.State;
    }
    if (isShareholder) {
      return EOwnership.Shareholders;
    }
    return EOwnership.Unknown;
  }

  private static getCeoAnnualPay(value: string | undefined) {
    if (!value) {
      return;
    }
    const regEx = /^\$(.+)m(.)*$/gi;
    const payMatch = regEx.exec(value);
    if (!payMatch) {
      return;
    }
    return parseFloat(payMatch[1]);
  }

  constructor(args: IPolluterArgs) {
    this.name = args.name;
    this.rank = args.rank;
    this.longitude = args.longitude;
    this.latitude = args.latitude;
    this.globalEmissions = args.globalEmissions;
    this.hq = args.hq;
    this.ceo = args.ceo || 'Unknown';
    this.fossilFuelProduction = args.fossilFuelProduction || NO_INFORMATTON;
    this.investmentInRenewables = args.investmentInRenewables || NO_INFORMATTON;
    this.globalEmissions1965to2017 = args.globalEmissions1965to2017 || NO_INFORMATTON;
    this.projectedIncreaseInProduction2018to2030 = args.projectedIncreaseInProduction2018to2030 || NO_INFORMATTON;
    this.projectedEmissions2018to2030 = args.projectedEmissions2018to2030 || NO_INFORMATTON;
    this.environmentalDisaster = args.environmentalDisaster;
    this.environmentalScandal = args.environmentalScandal;
    this.futureProjects = args.futureProjects || NO_INFORMATTON;
    this.revenue = args.revenue || NO_INFORMATTON;

    this.annualCeoPay = Polluter.getCeoAnnualPay(args.ceoAnnualPay);
    console.log(this.annualCeoPay);

    this.ownership = Polluter.getOwnership(args.ownership);
  }
}

export default Polluter;
