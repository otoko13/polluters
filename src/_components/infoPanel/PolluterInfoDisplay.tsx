import React from 'react';
import { IPolluter, EOwnership } from '../../model/Polluter';
import AssignmentIcon from '@material-ui/icons/Assignment';
import AttachMoneyIcon from '@material-ui/icons/AttachMoney';
import CloudIcon from '@material-ui/icons/Cloud';
import PersonIcon from '@material-ui/icons/Person';
import LocationOnIcon from '@material-ui/icons/LocationOn';
import LocalGasStationIcon from '@material-ui/icons/LocalGasStation';
import NextWeekIcon from '@material-ui/icons/NextWeek';
import PublicIcon from '@material-ui/icons/Public';
import TrendingUpIcon from '@material-ui/icons/TrendingUp';
import WarningIcon from '@material-ui/icons/Warning';
import PropertyDisplay from './PropertyDisplay';
import './polluterInfoDisplay.scss';
import OwnershipDisplay from './OwnershipDisplay';

export interface IPolluterInfoDisplayProps {
  selectedPolluter: IPolluter;
}

const neutralIconColor = 'grey';
const badIconColor = '#c45a68';
const goodIconColor = '#489c61';

const PolluterInfoDisplay = ({ selectedPolluter }: IPolluterInfoDisplayProps) =>
  (
    <div className="PolluterInfoDisplay">
      <PropertyDisplay title={"HQ"} icon={<LocationOnIcon fontSize="large" style={{color: neutralIconColor}} />}>
        {selectedPolluter.hq}
      </PropertyDisplay>

      <PropertyDisplay title={"Ownership"} icon={<AssignmentIcon fontSize="large" style={{color: neutralIconColor}} />}>
          <OwnershipDisplay ownership={selectedPolluter.ownership} />
      </PropertyDisplay>

      <PropertyDisplay title={"CEO"} icon={<PersonIcon fontSize="large" style={{color: neutralIconColor}} />}>
        {selectedPolluter.ceo} (annual pay: {selectedPolluter.annualCeoPay ? `$${selectedPolluter.annualCeoPay}m` : 'undisclosed'})
      </PropertyDisplay>

      <PropertyDisplay title={"Revenue"} icon={<AttachMoneyIcon fontSize="large" style={{color: neutralIconColor}} />}>
        {selectedPolluter.revenue}
      </PropertyDisplay>

      <PropertyDisplay title={"Fossil fuel production"} icon={<LocalGasStationIcon fontSize="large" style={{color: badIconColor}} />}>
        {selectedPolluter.fossilFuelProduction}
      </PropertyDisplay>

      <PropertyDisplay title={"Projected increase in production 2018-2030"} icon={<TrendingUpIcon fontSize="large" style={{color: badIconColor}} />}>
        {selectedPolluter.projectedIncreaseInProduction2018to2030}
      </PropertyDisplay>

      <PropertyDisplay title={"Global emissions 1965-2017"} icon={<CloudIcon fontSize="large" style={{color: badIconColor}} />}>
        {selectedPolluter.globalEmissions1965to2017}
      </PropertyDisplay>

      <PropertyDisplay title={"Projected emissions 2018-2030"} icon={<TrendingUpIcon fontSize="large" style={{color: badIconColor}} />}>
        {selectedPolluter.projectedIncreaseInProduction2018to2030}
      </PropertyDisplay>

      <PropertyDisplay title={"Future projects"} icon={<NextWeekIcon fontSize="large" style={{color: badIconColor}} />}>
        {selectedPolluter.futureProjects}
      </PropertyDisplay>

      {
        selectedPolluter.environmentalDisaster && (
          <PropertyDisplay title={"Environmental disaster"} icon={<WarningIcon fontSize="large" style={{color: badIconColor}} />}>
            {selectedPolluter.environmentalDisaster}
          </PropertyDisplay>
        )
      }

      {
        selectedPolluter.environmentalScandal && (
          <PropertyDisplay title={"Environmental scandal"} icon={<WarningIcon fontSize="large" style={{color: badIconColor}} />}>
            {selectedPolluter.environmentalScandal}
          </PropertyDisplay>
        )
      }

      <PropertyDisplay title={"Investment in renewables"} icon={<PublicIcon fontSize="large" style={{color: goodIconColor}} />}>
        {selectedPolluter.investmentInRenewables}
      </PropertyDisplay>
    </div>
  );

export default PolluterInfoDisplay;
