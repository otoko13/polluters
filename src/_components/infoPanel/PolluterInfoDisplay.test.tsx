import { render } from '@testing-library/react';
import * as React from 'react';
import PolluterInfoDisplay, { IPolluterInfoDisplayProps } from './PolluterInfoDisplay';
import { EOwnership, IPolluter } from '../../model/Polluter';

let props: IPolluterInfoDisplayProps;

beforeEach(() => {
  props = {
    selectedPolluter: {
      rank: 1,
      name: 'Test1',
      annualCeoPay: 25,
      ceo: 'Michael Buble',
      environmentalDisaster: 'Too many records',
      fossilFuelProduction: ' Lots of gas',
      futureProjects: 'Hopefully not',
      globalEmissions: 'Lots',
      globalEmissions1965to2017: 'Emissions for days',
      hq: 'London',
      investmentInRenewables: 'Unknown',
      ownership: EOwnership.StateAndShareholders,
      projectedEmissions2018to2030: 'Even more',
      projectedIncreaseInProduction2018to2030: '8 albums',
      revenue: '$2bn',
    } as IPolluter,
  };
});

describe('snapshots', () => {
  it('should look sensible', () => {
    const { container } = render(<PolluterInfoDisplay {...props} />);
    expect(container.firstChild).toMatchSnapshot('State');
  });
});
