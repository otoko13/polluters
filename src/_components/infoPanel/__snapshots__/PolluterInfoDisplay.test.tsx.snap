// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshots should look sensible: State 1`] = `
<div
  class="PolluterInfoDisplay"
>
  <div
    class="PropertyDisplay"
  >
    <div
      class="icon-container"
      title="HQ"
    >
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeLarge"
        focusable="false"
        style="color: grey;"
        viewBox="0 0 24 24"
      >
        <path
          d="M12 2C8.13 2 5 5.13 5 9c0 5.25 7 13 7 13s7-7.75 7-13c0-3.87-3.13-7-7-7zm0 9.5c-1.38 0-2.5-1.12-2.5-2.5s1.12-2.5 2.5-2.5 2.5 1.12 2.5 2.5-1.12 2.5-2.5 2.5z"
        />
      </svg>
    </div>
    <div
      class="content"
    >
      <div
        class="title"
      >
        HQ
      </div>
      London
    </div>
  </div>
  <div
    class="PropertyDisplay"
  >
    <div
      class="icon-container"
      title="Ownership"
    >
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeLarge"
        focusable="false"
        style="color: grey;"
        viewBox="0 0 24 24"
      >
        <path
          d="M19 3h-4.18C14.4 1.84 13.3 1 12 1c-1.3 0-2.4.84-2.82 2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zm-7 0c.55 0 1 .45 1 1s-.45 1-1 1-1-.45-1-1 .45-1 1-1zm2 14H7v-2h7v2zm3-4H7v-2h10v2zm0-4H7V7h10v2z"
        />
      </svg>
    </div>
    <div
      class="content"
    >
      <div
        class="title"
      >
        Ownership
      </div>
      <div
        class="OwnershipDisplay"
      >
        <div
          class="icons"
        >
          <span>
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M5.2496 8.0688l2.83-2.8268 14.134 14.15-2.83 2.8268zM9.4857 3.8272l2.828-2.8288 5.6576 5.656-2.828 2.8288zM.9989 12.3147l2.8284-2.8284L9.484 15.143l-2.8284 2.8284zM1 21h12v2H1z"
              />
            </svg>
             
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M16.67 13.13C18.04 14.06 19 15.32 19 17v3h4v-3c0-2.18-3.57-3.47-6.33-3.87z"
                fill-rule="evenodd"
              />
              <circle
                cx="9"
                cy="8"
                fill-rule="evenodd"
                r="4"
              />
              <path
                d="M15 12c2.21 0 4-1.79 4-4s-1.79-4-4-4c-.47 0-.91.1-1.33.24C14.5 5.27 15 6.58 15 8s-.5 2.73-1.33 3.76c.42.14.86.24 1.33.24zM9 13c-2.67 0-8 1.34-8 4v3h16v-3c0-2.66-5.33-4-8-4z"
                fill-rule="evenodd"
              />
            </svg>
          </span>
        </div>
        <div
          class="text"
        >
          State and shareholders
        </div>
      </div>
    </div>
  </div>
  <div
    class="PropertyDisplay"
  >
    <div
      class="icon-container"
      title="CEO"
    >
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeLarge"
        focusable="false"
        style="color: grey;"
        viewBox="0 0 24 24"
      >
        <path
          d="M12 12c2.21 0 4-1.79 4-4s-1.79-4-4-4-4 1.79-4 4 1.79 4 4 4zm0 2c-2.67 0-8 1.34-8 4v2h16v-2c0-2.66-5.33-4-8-4z"
        />
      </svg>
    </div>
    <div
      class="content"
    >
      <div
        class="title"
      >
        CEO
      </div>
      Michael Buble
       (annual pay: 
      $25m
      )
    </div>
  </div>
  <div
    class="PropertyDisplay"
  >
    <div
      class="icon-container"
      title="Revenue"
    >
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeLarge"
        focusable="false"
        style="color: grey;"
        viewBox="0 0 24 24"
      >
        <path
          d="M11.8 10.9c-2.27-.59-3-1.2-3-2.15 0-1.09 1.01-1.85 2.7-1.85 1.78 0 2.44.85 2.5 2.1h2.21c-.07-1.72-1.12-3.3-3.21-3.81V3h-3v2.16c-1.94.42-3.5 1.68-3.5 3.61 0 2.31 1.91 3.46 4.7 4.13 2.5.6 3 1.48 3 2.41 0 .69-.49 1.79-2.7 1.79-2.06 0-2.87-.92-2.98-2.1h-2.2c.12 2.19 1.76 3.42 3.68 3.83V21h3v-2.15c1.95-.37 3.5-1.5 3.5-3.55 0-2.84-2.43-3.81-4.7-4.4z"
        />
      </svg>
    </div>
    <div
      class="content"
    >
      <div
        class="title"
      >
        Revenue
      </div>
      $2bn
    </div>
  </div>
  <div
    class="PropertyDisplay"
  >
    <div
      class="icon-container"
      title="Fossil fuel production"
    >
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeLarge"
        focusable="false"
        style="color: rgb(196, 90, 104);"
        viewBox="0 0 24 24"
      >
        <path
          d="M19.77 7.23l.01-.01-3.72-3.72L15 4.56l2.11 2.11c-.94.36-1.61 1.26-1.61 2.33 0 1.38 1.12 2.5 2.5 2.5.36 0 .69-.08 1-.21v7.21c0 .55-.45 1-1 1s-1-.45-1-1V14c0-1.1-.9-2-2-2h-1V5c0-1.1-.9-2-2-2H6c-1.1 0-2 .9-2 2v16h10v-7.5h1.5v5c0 1.38 1.12 2.5 2.5 2.5s2.5-1.12 2.5-2.5V9c0-.69-.28-1.32-.73-1.77zM12 10H6V5h6v5zm6 0c-.55 0-1-.45-1-1s.45-1 1-1 1 .45 1 1-.45 1-1 1z"
        />
      </svg>
    </div>
    <div
      class="content"
    >
      <div
        class="title"
      >
        Fossil fuel production
      </div>
       Lots of gas
    </div>
  </div>
  <div
    class="PropertyDisplay"
  >
    <div
      class="icon-container"
      title="Projected increase in production 2018-2030"
    >
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeLarge"
        focusable="false"
        style="color: rgb(196, 90, 104);"
        viewBox="0 0 24 24"
      >
        <path
          d="M16 6l2.29 2.29-4.88 4.88-4-4L2 16.59 3.41 18l6-6 4 4 6.3-6.29L22 12V6z"
        />
      </svg>
    </div>
    <div
      class="content"
    >
      <div
        class="title"
      >
        Projected increase in production 2018-2030
      </div>
      8 albums
    </div>
  </div>
  <div
    class="PropertyDisplay"
  >
    <div
      class="icon-container"
      title="Global emissions 1965-2017"
    >
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeLarge"
        focusable="false"
        style="color: rgb(196, 90, 104);"
        viewBox="0 0 24 24"
      >
        <path
          d="M19.35 10.04C18.67 6.59 15.64 4 12 4 9.11 4 6.6 5.64 5.35 8.04 2.34 8.36 0 10.91 0 14c0 3.31 2.69 6 6 6h13c2.76 0 5-2.24 5-5 0-2.64-2.05-4.78-4.65-4.96z"
        />
      </svg>
    </div>
    <div
      class="content"
    >
      <div
        class="title"
      >
        Global emissions 1965-2017
      </div>
      Emissions for days
    </div>
  </div>
  <div
    class="PropertyDisplay"
  >
    <div
      class="icon-container"
      title="Projected emissions 2018-2030"
    >
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeLarge"
        focusable="false"
        style="color: rgb(196, 90, 104);"
        viewBox="0 0 24 24"
      >
        <path
          d="M16 6l2.29 2.29-4.88 4.88-4-4L2 16.59 3.41 18l6-6 4 4 6.3-6.29L22 12V6z"
        />
      </svg>
    </div>
    <div
      class="content"
    >
      <div
        class="title"
      >
        Projected emissions 2018-2030
      </div>
      8 albums
    </div>
  </div>
  <div
    class="PropertyDisplay"
  >
    <div
      class="icon-container"
      title="Future projects"
    >
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeLarge"
        focusable="false"
        style="color: rgb(196, 90, 104);"
        viewBox="0 0 24 24"
      >
        <path
          d="M20 7h-4V5c0-.55-.22-1.05-.59-1.41C15.05 3.22 14.55 3 14 3h-4c-1.1 0-2 .9-2 2v2H4c-1.1 0-2 .9-2 2v11c0 1.1.9 2 2 2h16c1.1 0 2-.9 2-2V9c0-1.1-.9-2-2-2zM10 5h4v2h-4V5zm1 13.5l-1-1 3-3-3-3 1-1 4 4-4 4z"
        />
      </svg>
    </div>
    <div
      class="content"
    >
      <div
        class="title"
      >
        Future projects
      </div>
      Hopefully not
    </div>
  </div>
  <div
    class="PropertyDisplay"
  >
    <div
      class="icon-container"
      title="Environmental disaster"
    >
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeLarge"
        focusable="false"
        style="color: rgb(196, 90, 104);"
        viewBox="0 0 24 24"
      >
        <path
          d="M1 21h22L12 2 1 21zm12-3h-2v-2h2v2zm0-4h-2v-4h2v4z"
        />
      </svg>
    </div>
    <div
      class="content"
    >
      <div
        class="title"
      >
        Environmental disaster
      </div>
      Too many records
    </div>
  </div>
  <div
    class="PropertyDisplay"
  >
    <div
      class="icon-container"
      title="Investment in renewables"
    >
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeLarge"
        focusable="false"
        style="color: rgb(72, 156, 97);"
        viewBox="0 0 24 24"
      >
        <path
          d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm-1 17.93c-3.95-.49-7-3.85-7-7.93 0-.62.08-1.21.21-1.79L9 15v1c0 1.1.9 2 2 2v1.93zm6.9-2.54c-.26-.81-1-1.39-1.9-1.39h-1v-3c0-.55-.45-1-1-1H8v-2h2c.55 0 1-.45 1-1V7h2c1.1 0 2-.9 2-2v-.41c2.93 1.19 5 4.06 5 7.41 0 2.08-.8 3.97-2.1 5.39z"
        />
      </svg>
    </div>
    <div
      class="content"
    >
      <div
        class="title"
      >
        Investment in renewables
      </div>
      Unknown
    </div>
  </div>
</div>
`;
